bug 1113260 - use django-ratelimit for ip bans

groovecoder opened the pull request:
**DNM** Sanity-check PR. Here's the approach so far:

* Adding a new `IPBan` model to `kuma.core` (like `UserBan` in `kuma.users`)
* Decorate views with `ratelimit` of a new `limit_banned_ip_to_0` util function

Pro: Simpler code that uses `django-ratelimit` (a well-supported popular library) as its designed

Con: Have to decorate each view (or add mixin to CBV's)

Overall I'm liking this approach more than `django-banish` or any others we've tried so far.

groovecoder commented:
Can I get a sanity check from @jezdez, @ubernostrum, @robhudson ... or even @jsocol? :wink:

jsocol commented on kuma/core/utils.py:
```diff
@@ -284,3 +286,11 @@ def chord_flow(pre_task, tasks, post_task):
+
+def limit_banned_ip_to_0(group, request):
+    try:
+        IPBan.objects.get(ip=get_ip(request))
```
If this ever gets applied to more-frequent endpoints, I'd put some layer of caching here.

jsocol commented:
This is exactly the kind of thing I hoped people would build when I made the decorator accept callables!

Middleware might be more effective if you are willing to be coarser (i.e. block _all_ requests, or all unsafe HTTP methods, from a given IP) in exchange for writing way less code. That said, putting things that can fail (like DB/cache lookups) in middleware is discouraged because the error handling is rough, so it's important to think about what those failure cases look like.

The nice thing about doing this at the application layer is that the application can control things relatively easily. The downside is that the application still consumes (some) resources serving banned users, so if you are willing to be coarser, something at a higher level, like fail2ban, may also be an option.

Finally, I made some specific comments, because I miss you folks.

jsocol commented on kuma/core/managers.py:
```diff
@@ -241,6 +243,13 @@ def resolve_allowed_tags(model_obj, tags_curr, tags_new,
 
 
+class IPBanManager(models.Manager):
+    def delete_old(self, days=30):
```
IP addresses aren't _that_ stable, 30 days seems high as a default.

jsocol commented on kuma/core/managers.py:
```diff
@@ -241,6 +243,13 @@ def resolve_allowed_tags(model_obj, tags_curr, tags_new,
+    def delete_old(self, days=30):
+        cutoff_date = date.today() - timedelta(days=days)
+        old_ip_bans = self.get_query_set().filter(created__lte=cutoff_date)
+        old_ip_bans.delete()
```
I've taken to adding `deleted = models.DateTimeField(null=True,blank=True)` instead of actually deleting things, so there's some history. Then instead of actually deleting, you can make the model's/manager's `delete()` method just set the `deleted` attribute, and since you have a manager here anyway, maybe make an `active()` that does `super().filter(deleted__isnull=True)`.

robhudson commented on kuma/core/models.py:
```diff
@@ -34,6 +37,14 @@ def update(self, **kw):
 
+class IPBan(models.Model):
+    ip = models.CharField(max_length=40, editable=False, db_index=True,
+                          blank=True, null=True, unique=True)
```
Would using https://docs.djangoproject.com/en/1.7/ref/models/fields/#genericipaddressfield have any benefit?

jsocol commented on kuma/core/models.py:
```diff
@@ -34,6 +37,14 @@ def update(self, **kw):
+class IPBan(models.Model):
+    ip = models.CharField(max_length=40, editable=False, db_index=True,
+                          blank=True, null=True, unique=True)
+    created = models.DateTimeField(default=datetime.now, db_index=True)
```
Re: the caching comment, I've been doing something like:

```python
@classmethod
def get(cls, some_natural_key):
    obj = cache.get(cls._cache_key(some_natural_key))
    if obj == EMPTY_SENTINEL:  # e.g. '-'
        raise cls.DoesNotExist  # or return None
    # ... etc
```

IP address, and `deleted__isnull` make a pretty natural key here.

groovecoder commented:
Updated to fully functional code. I'm sticking with the single decorator because this is first and foremost meant to implement https://bugzilla.mozilla.org/show_bug.cgi?id=1113260. We may add middleware when we go on to implement automatic banning & limiting in https://bugzilla.mozilla.org/show_bug.cgi?id=1122658.

Still needs tests ... that's next.

groovecoder commented:
Okay, this is ready with tests and everything. An outstanding question for @jezdez @ubernostrum - how should we clean out the db tables from `django-banish`? Data migration in `kuma.core`? Or manually?

jezdez commented on kuma/core/managers.py:
```diff
@@ -241,6 +243,16 @@ def resolve_allowed_tags(model_obj, tags_curr, tags_new,
+
+    def delete_old(self, days=30):
+        cutoff_date = date.today() - timedelta(days=days)
+        old_ip_bans = self.get_query_set().filter(created__lte=cutoff_date)
```
`self.filter(..)` is enough, the call to `get_query_set` is implied when using `self`

jezdez commented on kuma/core/managers.py:
```diff
@@ -241,6 +243,16 @@ def resolve_allowed_tags(model_obj, tags_curr, tags_new,
 
+class IPBanManager(models.Manager):
+    def active(self, ip):
+        return super(IPBanManager, self).filter(ip=ip, deleted__isnull=True)
```
Just use `self.filter` here, no need to use `super`

jezdez commented on kuma/core/models.py:
```diff
@@ -34,6 +37,21 @@ def update(self, **kw):
 
+class IPBan(models.Model):
+    ip = models.GenericIPAddressField()
+    created = models.DateTimeField(default=datetime.now, db_index=True)
```
Let's use `django.utils.timezone.now` for the default in preparation to make kuma timezone aware (a feature in current Django)

jezdez commented on kuma/core/models.py:
```diff
@@ -34,6 +37,21 @@ def update(self, **kw):
+    objects = IPBanManager()
+
+    def delete(self, *args, **kwargs):
+        self.deleted = datetime.now()
```
Use `timezone.now` as well

jezdez commented on kuma/core/models.py:
```diff
@@ -34,6 +37,21 @@ def update(self, **kw):
+
+    def delete(self, *args, **kwargs):
+        self.deleted = datetime.now()
+        self.save()
```
This mean's we can never delete that db entry, which is not a good pattern. There are apps out there implementing that logical delete pattern much better, I'd strongly recommend using that instead.

groovecoder replied:
The 2 simple pluggable apps I've found so far are:

https://github.com/paltman/django-logicaldelete
https://github.com/hearsaycorp/django-livefield

[Others](https://www.djangopackages.com/grids/g/deletion/) seem like overkill. I'll check them out next week.

jezdez commented on kuma/core/tasks.py:
```diff
@@ -51,3 +52,8 @@ def clean_sessions():
+
+
+@task
+def delete_old_ip_bans(immediate=False, days=30):
```
Unused `immediate` param

jezdez commented on kuma/core/tasks.py:
```diff
@@ -7,6 +7,7 @@
 import constance.config
 
 from kuma.core.cache import memcache
+from .models import IPBan
```
Please add empty line before local import.

jezdez commented on kuma/core/tests/test_models.py:
```diff
@@ -0,0 +1,36 @@
+
+from django.test import TestCase
+
+from kuma.core.models import IPBan
```
Please use local import, `from ..models import IPBan`

jezdez commented on kuma/core/tests/test_models.py:
```diff
@@ -0,0 +1,36 @@
+class RevisionIPTests(TestCase):
+    def test_delete_older_than_default_30_days(self):
+        old_date = date.today() - timedelta(days=31)
+        IPBan.objects.create(ip='127.0.0.1', created=old_date).save()
```
`create` calls `save` automatically, no need for duplication

jezdez commented on kuma/core/tests/test_models.py:
```diff
@@ -0,0 +1,36 @@
+
+    def test_delete_older_than_days_argument(self):
+        ban_date = date.today() - timedelta(days=5)
+        IPBan(ip='127.0.0.1', created=ban_date).save()
```
either use `objects.create` or `save`

jezdez commented on kuma/core/tests/test_models.py:
```diff
@@ -0,0 +1,36 @@
+    def test_delete_older_than_default_30_days(self):
+        old_date = date.today() - timedelta(days=31)
+        IPBan.objects.create(ip='127.0.0.1', created=old_date).save()
+        eq_(1, IPBan.objects.all().count())
```
`all()` is not needed here, it's implied when running `count()`, also happens in lines below

jezdez commented on kuma/core/utils.py:
```diff
@@ -284,3 +286,9 @@ def chord_flow(pre_task, tasks, post_task):
+
+
+def limit_banned_ip_to_0(group, request):
+    if IPBan.objects.active(ip=get_ip(request)).count() > 0:
```
Please use `exists` instead of `count() > 0` since `exists` has a slightly improved internal cache

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -547,6 +548,34 @@ def test_banned_users(self):
+        super(BannedIPTests, self).setUp()
+        self.ip = '127.0.0.1'
+        self.ip_ban = IPBan.objects.create(ip=self.ip)
+        self.d, r = doc_rev()
```
please don't use single character variable names

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -547,6 +548,34 @@ def test_banned_users(self):
+
+    def test_banned_ip_cant_get_edit(self):
+        self.client.login(username='testuser', password='testpass')
+        resp = self.client.get(self.edit_url, REMOTE_ADDR=self.ip)
```
s/resp/response/g

jezdez replied:
as well as below

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -547,6 +548,34 @@ def test_banned_users(self):
+        eq_(403, resp.status_code)
+
+    def test_banned_ip_can_still_get_articles(self):
+        resp = self.client.get(reverse('wiki.document',
```
Please use `self.document.get_absolute_url`

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -2354,6 +2383,9 @@ def test_store_revision_ip(self):
         rev_ip = RevisionIP.objects.get(revision=rev)
         eq_('127.0.0.1', rev_ip.ip)
 
+    def test_banned_ip(self):
```
Something you want to fill?

jezdez commented on kuma/core/utils.py:
```diff
@@ -284,3 +286,9 @@ def chord_flow(pre_task, tasks, post_task):
+def limit_banned_ip_to_0(group, request):
+    if IPBan.objects.active(ip=get_ip(request)).count() > 0:
+        return "0/s"
+    return "60/m"
```
This *definitely* needs caching, no way around it. This is going to be called *a lot* and no matter how optimized the MySQL cluster is, counts and checks are slow. Merge #3115 and add a django-cacheback job to handle the cache invalidation nicely.

jezdez commented:
Ok, I think we're on the right track, but this needs some updates, mostly code smell but one biggy: caching. See #3115 for a new way to handle cache refills via Celery nicely that should be used here as well.

groovecoder commented:
Updated with all the smaller updates. Going to check out https://github.com/mozilla/kuma/pull/3115 caching strategy before adding the caching here.

groovecoder commented:
Added https://github.com/groovecoder/kuma/commit/53b749611daef03f68ca2e9c730e57aec55a11ac with a `django-cacheback` `IPBanJob`. It also includes some of the same changes in #3115, so we should finish and merge that first. Then I can rebase this PR.

jezdez commented on kuma/core/jobs.py:
```diff
@@ -18,3 +18,17 @@ def hash(self, value):
+        return "60/m"
+
+    def empty(self):
+        return []
```
I don't think it should return a list here since you're expecting a string in fetch. the docstring says:

        Return the appropriate value for a cache MISS (and when we defer the
        repopulation of the cache)

So probably something like `"60/m"` as well?

jezdez commented on kuma/core/models.py:
```diff
@@ -34,6 +38,22 @@ def update(self, **kw):
+    def delete(self, *args, **kwargs):
+        self.deleted = timezone.now()
+        self.save()
+        IPBanJob.delete(self.ip)
```
This needs to be using an instance of a job, so `IPBanJob().delete(self.ip)`. That said, I think we should use the `refresh` method since that not only deletes the cache entry but fills the cache synchronously again right away and doesn't wait till the next request to fill it.

jezdez commented on kuma/core/utils.py:
```diff
@@ -19,7 +19,9 @@
 
 from taggit.utils import split_strip
 
+from kuma.actioncounters.utils import get_ip
```
Let's move that function into core.

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -547,6 +548,39 @@ def test_banned_users(self):
+        self.client.login(username='testuser', password='testpass')
+        response = self.client.get(self.edit_url, REMOTE_ADDR=self.ip)
+        eq_(403, response.status_code)
+        cache.clear()
```
Move the call to `cache.clear` into the `setUp` method instead.

groovecoder commented:
Updated.

I moved `cache.clear` to `tearDown` because it's cleaning up the cache for tests **following** the `BannedIPTests` tests - including tests in other classes.

jezdez commented on .gitmodules:
```diff
@@ -148,3 +145,9 @@
+[submodule "vendor/src/django-ratelimit"]
+	path = vendor/src/django-ratelimit
+	url = git://github.com/jsocol/django-ratelimit.git
+[submodule "vendor/src/django-cacheback"]
```
Hm, why is that adding that twice?

jezdez commented on kuma/core/models.py:
```diff
@@ -34,6 +38,22 @@ def update(self, **kw):
+    objects = IPBanManager()
+
+    def delete(self, *args, **kwargs):
+        self.deleted = timezone.now()
```
What's the verdict about logicial deletes btw?

jezdez replied:
I mean apps.

groovecoder replied:
All the ones I saw in https://www.djangopackages.com/grids/g/deletion/ looked either too heavyweight or unstable/un-maintained. :frowning: Unless one of them looks right to you?

jezdez replied:
Hm, I haven't looked for real. If that's not the way you wanna go now (we can do that later (soon)), can you add the deleted field to the admin list display?

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -547,6 +548,40 @@ def test_banned_users(self):
+        self.doc, rev = doc_rev()
+        self.edit_url = reverse('wiki.edit_document',
+                                args=[self.doc.full_path]
+                               )
```
Move the closing bracket up one line

jezdez commented on kuma/wiki/tests/test_views.py:
```diff
@@ -547,6 +548,40 @@ def test_banned_users(self):
+    def test_banned_ip_can_still_get_articles(self):
+        response = self.client.get(self.doc.get_absolute_url(),
+                                   REMOTE_ADDR=self.ip
+                                  )
```
Bracket up one line

jezdez commented:
LGTM other those last minor issues r+

groovecoder commented:
Updated. Waiting on Travis ...

groovecoder commented:
Added the custom `list_display` for the admin site.

groovecoder commented:
Read for final review & merge ...

jezdez commented on kuma/contentflagging/utils.py:
```diff
@@ -1,40 +1,6 @@
-            pass
-
-    return ip_address
+from kuma.core.utils import get_ip
```
Yay for code deduping!

jezdez commented:
r+
